johnzon-core/src/main/java/org/apache/johnzon/core/JohnzonJsonParserImpl.java

```diff
@@ -134,18 +140,57 @@ public void skipArray() {
 
     @Override
     public Stream<JsonValue> getArrayStream() {
-        //X TODO this implementation is very simplistic
-        //X I find it unintuitive what the spec intends here
-        //X we probably need to improve this
-        return getArray().stream();
+        Event current = current();
+        if (current != Event.START_ARRAY) {
+            throw new IllegalStateException(current + " doesn't support getArrayStream()");
+        }
+
+        Spliterator<JsonValue> arraySpliterator = new Spliterators.AbstractSpliterator<JsonValue>(
```

Comment:
What about making it an inner private class reused in both locations (private static class)?

Reply:
The generic arguments and the `tryAdvance()` implementations are different, so what could be reused? As far as I can see, the only thing duplicated between the two Spliterator implementations are the arguments passed to the superclass constructor. I can extract that to a helper class if you'd like, but it feels overkill to me.

Reply:
if you think you need 2 inner classes it is fine for me too, what I want to ensure is that it is not referencing $0 (this).
it is also fine ot inject the impl of tryAdvance as a lambda (the compiler will make it right normally) so maybe something along:

> new SimpleSplitIterator(action -> {  Event next = next(); ..... });

this enables to factorize the config and avoids the $0 ref.

Reply:
It needs to reference the enclosing parser instance, though, because it is calling methods on that instance like `next()`. If `tryAdvance` is factored out into a lambda, it will still capture `this`. As far as I can tell, a lambda won't help anything in this situation. It either needs to stay a nonstatic class like it is now, or it can be turned into a static class, but instances of that static class will need to explicitly reference the parser they came from.

Reply:
Pass parser as a param ;)
Lambdas have 3 conversions so if well refactored it solves it.
You can use javap to ensure you reached that goal

Reply:
I've pushed some changes. Is this what you had in mind?